pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1334,15 +1334,26 @@ func (c *PodAffinityChecker) satisfiesExistingPodsAntiAffinity(pod *v1.Pod, meta
 //  topology of all the "terms" for the given "pod".
 func (c *PodAffinityChecker) nodeMatchesAllTopologyTerms(pod *v1.Pod, topologyPairs *topologyPairsMaps, nodeInfo *schedulercache.NodeInfo, terms []v1.PodAffinityTerm) bool {
 	node := nodeInfo.Node()
-	for _, term := range terms {
-		if topologyValue, ok := node.Labels[term.TopologyKey]; ok {
-			pair := topologyPair{key: term.TopologyKey, value: topologyValue}
-			if _, ok := topologyPairs.topologyPairToPods[pair]; !ok {
-				return false
-			}
-		} else {
+OUTERLOOP:
+	for idx, term := range terms {
+		topologyValue, ok := node.Labels[term.TopologyKey]
+		if !ok {
 			return false
 		}
+
+		pair := topologyPair{key: term.TopologyKey, value: topologyValue}
+		podSet, ok := topologyPairs.topologyPairToPods[pair]
+		if !ok {
+			return false
+		}
+		// "idx" should be found in "podSet[<any pod>]"
+		for _, bitwisedIdx := range podSet {
```

Comment:
Why do we need this check? I appears to me that when all topologyPairs of terms exist in the topologyPairs.topologyPairToPods, affinity is satisfied.

Reply:
It's because right now `topologyPairs.topologyPairToPods[pair]`, which is `podSet map[*v1.Pod]uint64`, maintain a relation of "which pod matches on which index of the terms" - it's like a "contribution" to the final match on all terms of PodAffinity. Additionally, two pods in different nodes can "contribute" to the same term, and failed on other term(s). So counting numbers doesn't help, here I used the bitwised ORed value.

So in this method, the logic is: only when all terms's index can be found in some pod, we treat current node as a match.

("some pod" is not necessarily on current node, but it's in the same topology domain of the node)

Reply:
@bsalamat latest commit 2e88364e34157201d75b7931e96ca6d7964d6c14 added a testcase which failed if we don't have bitmap check.

Reply:
But that case can be resolved by "merging" mergeable terms (terms which has the same topologyKey).

Let me think more to see if there are any other corner cases.

Reply:
> @bsalamat latest commit 2e88364 added a testcase which failed if we don't have bitmap check.

Yes, you are right. In that case, we should replace the bit map with an actual map to store the indeces of terms. I understand that probably no one adds more than 64 terms to an affinity rule, but we cannot have that restriction in our implementation.

Reply:
> but we cannot have that restriction in our implementation

Understood.

Basically here is what I'm thinking of at this point:

- option1: continue current solution, but need extra efforts on server validation (maybe as well as kubectl validation), and also docs
- option2: don't use bitmap, revert back to a simple struct{}. And when computing topologyPairToPods for PodAffinity, perform a "dry-run merge" on terms which has the same topology. Say only when a "merged" term matches, we add it to topologyPairToPods.

Reply:
option3: use an actual map/set to store the indexes of matched terms instead of the current bitmap. ;-)

Reply:
sure, option3 would be the most reasonable one.

I can make podSet from current `map[*v1.Pod]struct{}` to `map[*v1.Pod]intSet` (`type intSet map[int]struct{}`).

Reply:
Yes, that would be better IMO.